aws/resource_aws_lb_listener_rule.go

```diff
@@ -283,18 +284,157 @@ func resourceAwsLbbListenerRule() *schema.Resource {
+								"http-request-method",
+								"path-pattern",
+								"query-string",
+								"source-ip",
+							}, true),
 						},
-						"values": {
+						"host_header": {
 							Type:     schema.TypeList,
 							MaxItems: 1,
-							Elem:     &schema.Schema{Type: schema.TypeString},
 							Optional: true,
+							// Deprecated: Remove Computed: true in next major version of the provider
+							Computed:         true,
+							DiffSuppressFunc: suppressIfConditionFieldNotIn([]string{"host-header"}),
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"values": {
+										Type: schema.TypeList,
+										Elem: &schema.Schema{
+											Type:         schema.TypeString,
+											ValidateFunc: validation.StringLenBetween(1, 128),
+										},
+										Required: true,
+									},
+								},
+							},
+						},
+						"http_header": {
+							Type:             schema.TypeList,
+							MaxItems:         1,
+							Optional:         true,
+							DiffSuppressFunc: suppressIfConditionFieldNotIn([]string{"http-header"}),
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"http_header_name": {
+										Type:         schema.TypeString,
+										Required:     true,
+										ValidateFunc: validation.StringLenBetween(1, 40),
+									},
+									"values": {
+										Type: schema.TypeList,
+										Elem: &schema.Schema{
+											Type:         schema.TypeString,
+											ValidateFunc: validation.StringLenBetween(1, 128),
+										},
+										Required: true,
+									},
+								},
+							},
+						},
+						"http_request_method": {
+							Type:             schema.TypeList,
+							MaxItems:         1,
+							Optional:         true,
+							DiffSuppressFunc: suppressIfConditionFieldNotIn([]string{"http-request-method"}),
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"values": {
+										Type: schema.TypeList,
+										Elem: &schema.Schema{
+											Type:         schema.TypeString,
+											ValidateFunc: validation.StringMatch(regexp.MustCompile(`^[A-Za-z-_]{1,40}$`), ""),
+										},
+										Required: true,
+									},
+								},
+							},
+						},
+						"path_pattern": {
+							Type:     schema.TypeList,
+							MaxItems: 1,
+							Optional: true,
+							// Deprecated: Remove Computed: true in next major version of the provider
+							Computed:         true,
+							DiffSuppressFunc: suppressIfConditionFieldNotIn([]string{"path-pattern"}),
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"values": {
+										Type: schema.TypeList,
+										Elem: &schema.Schema{
+											Type:         schema.TypeString,
+											ValidateFunc: validation.StringLenBetween(1, 128),
+										},
+										Required: true,
+									},
+								},
+							},
+						},
+						"query_string": {
+							Type:             schema.TypeList,
+							MaxItems:         1,
```

Comment:
This is for a single `query_string` block inside the condition resource. Within that block you specify a single `values` list of key/value pairs. This is a thin wrapper around the AWS API and is consistent pattern for all these new blocks. Admittedly, it works better for the other types. https://docs.aws.amazon.com/elasticloadbalancing/latest/APIReference/API_RuleCondition.html

Your two examples would be written as:
```HCL
resource "aws_lb_listener_rule" "test" {
  listener_arn = "${aws_lb_listener.this.arn}"

  action {
    type = "fixed-response"

    fixed_response {
      status_code  = "503"
      content_type = "text/plain"
    }
  }

  condition {
    field = "query-string"

    query_string {
      values = [
        {
          key   = "foobar"
          value = "baz"
        },
      ]
    }
  }

  condition {
    field = "path-pattern"

    path_pattern {
      values = ["1"]
    }
  }

  condition {
    field = "query-string"

    query_string {
      values = [
        {
          key   = "foo"
          value = "bar"
        },
      ]
    }
  }

  condition {
    field = "query-string"

    query_string {
      values = [
        {
          key   = "hello"
          value = "world"
        },
      ]
    }
  }

  condition {
    field = "host-header"

    host_header {
      values = ["example.com"]
    }
  }
}
```

```HCL
resource "aws_lb_listener_rule" "test" {
  listener_arn = "${aws_lb_listener.this.arn}"

  action {
    ...
  }

  condition {
    field = "query-string"

    query_string {
      values = [
        {
          key   = "foo"
          value = "bar"
        },
        {
          key   = "foobar"
          value = "baz"
        },
        {
          key   = "hello"
          value = "world"
        },
      ]
    }
  }
}
```

Coming back to look at this now, it is rather verbose. Maybe something could be done to flatten it down. 
```HCL
condition {
  field = "query-string"

  query_string {
    values = {
      foo    = "bar"
      foobar = "baz"
      hello  = "world"
    }
  }
}
```

Or to an extreme, like `tags`. This adds quite an abstraction over the API which could potentially change again in the future. It also breaks pattern with the other blocks:
```HCL
condition {
  field = "query-string"

  query_string {
    foo    = "bar"
    foobar = "baz"
    hello  = "world"
  }
}
```

Reply:
Is there value here in supporting multiple `query_string` blocks within a single condition? How would that be handled internally by terraform? We have to flatten into a single `QueryStringConditionConfig` object to the AWS API and then somehow split it up correctly when refreshing.

Reply:
> Is there value here in supporting multiple `query_string` blocks within a single condition?
> There's also the issue of assumptions people make. With a single allowed block that maps nicely to the rule line of OR strings inside AWS console. But if you have multiple query_string blocks within a condition block are you expecting them to get flattened to a single OR list or be separate AND rule conditions?

I think there is: it _seems_ to me multiple blocks are logical AND, while a single block with multiple values makes them OR.
It may needs some testing though, which, if you can do would be good, or I will do rather soon.

> In the latter we can easily take each query_string block to be different QueryStringConditionConfig objects but refreshing is hard. Do we assume a single query-string condition block with multiple query_string blocks? What happens if the user has multiple query-string condition blocks?

I agree in that case it starts to be a schema mess, which is not easily solvable. I'll think about it more

Reply:
I was just updating the documentation and it reminded me of the features here. In `query-string`:
- the key is optional
- key can be repeated with different values

A simple map will not be suitable.

I'm not sure how to reformat this to still support those conditions:
```HCL
condition {
  field = "query-string"

  query_string {
    values = [
      {
        value = "fuzz"
      },
      {
        key   = "foo"
        value = "bar"
      },
      {
        key   = "foo"
        value = "baz"
      },
    ]
  }
}
```

Maybe:
```HCL
query_string {
  ""  = ["fuzz"]
  foo = ["bar", "baz"]
}
```
Although this is abstracting further from the original API and may make it difficult to work with.

Reply:
I think I finally understand what you were saying in the documentation section comment due to the quirk of how terraform works:

```HCL
resource "aws_lb_listener_rule" "test" {
  listener_arn = "${aws_lb_listener.this.arn}"

  action {
    type = "fixed-response"

    fixed_response {
      status_code  = "503"
      content_type = "text/plain"
    }
  }

  condition {
    field = "query-string"

    query_string {
      values {
        key   = "foobar"
        value = "baz"
      }
    }
  }

  condition {
    field = "path-pattern"

    path_pattern {
      values = ["1"]
    }
  }

  condition {
    field = "query-string"

    query_string {
      values {
        key   = "foo"
        value = "bar"
      }
    }
  }

  condition {
    field = "query-string"

    query_string {
      values {
        key   = "hello"
        value = "world"
      }
    }
  }

  condition {
    field = "host-header"

    host_header {
      values = ["example.com"]
    }
  }
}

resource "aws_lb_listener_rule" "test2" {
  listener_arn = "${aws_lb_listener.this.arn}"

  action {
    type = "fixed-response"

    fixed_response {
      status_code  = "503"
      content_type = "text/plain"
    }
  }

  condition {
    field = "query-string"

    query_string {
      values {
        key   = "foo"
        value = "bar"
      }

      values {
        key   = "foobar"
        value = "baz"
      }

      values {
        key   = "hello"
        value = "world"
      }
    }
  }
}
```

Simple map wouldn't work as key can be repeated with different values. It's also optional:
```hcl
query_string {
  values {
    value = "bar"
  }

  values {
    key   = "foo"
    value = "baz"
  }

  values {
    key   = "foo"
    value = "buzz"
  }
}